### DISCUSSION TITLE
Site replicate not success

### DISCUSSION THREAD

**Discussion by kittisuw** [minio/minio#15101] [Q&A] [answered]
I need to set Site replicate 2 Minio server as manual https://www.digitalocean.com/community/tutorials/how-to-set-up-minio-object-storage-server-in-standalone-mode-on-ubuntu-20-04 after I set to 2 servers I using the command below on server1 and can put file to both server by mcli success
```
mcli --insecure admin info poc-minio-01
●  10.50.128.8:9000
   Uptime: 34 minutes
   Version: 2022-06-11T19:55:32Z
   Network: 1/1 OK
   Drives: 1/1 OK
   Pool: 1st

1 drive online, 0 drives offline

mcli --insecure admin info poc-minio-02
●  10.50.128.9:9000
   Uptime: 12 minutes
   Version: 2022-06-11T19:55:32Z
   Network: 1/1 OK
   Drives: 1/1 OK
   Pool: 1st

mcli --insecure admin replicate add poc-minio-01 poc-minio-02
```
output
```shell
mcli: <ERROR> Unable to add sites for replication. Error received when contacting a peer site (unable to fetch server info for poc-minio-02: Get "https://10.50.128.9:9000/minio/admin/v3/info": x509: certificate signed by unknown authority).
```
after this I was fix allow all IP of 2 node in step certgen on both server
```shell
sudo certgen -host 40.65.137.16,10.50.128.8,52.148.71.42,10.50.128.9
Created a new certificate 'public.crt', 'private.key' valid for the following names 📜
 - "40.65.137.16"
 - "10.50.128.8"
 - "52.148.71.42"
 - "10.50.128.9"
```
but error still the same

**Comment by harshavardhana** [comment, marked as answer]
> mcli: <ERROR> Unable to add sites for replication. Error received when contacting a peer site (unable to fetch server info for poc-minio-02: Get "https://10.50.128.9:9000/minio/admin/v3/info": x509: certificate signed by unknown authority).

Put your public.crt as a trust under `/.minio/certs/CAs` folder?

**Reply by kittisuw** [reply]
Here is minio config 

File :/etc/default/minio
```
MINIO_VOLUMES="/mnt/data"
MINIO_OPTS="--certs-dir /home/sammy/.minio/certs --console-address :9001"
MINIO_ROOT_USER=minioadmin
MINIO_ROOT_PASSWORD=minioadmin
```
and i put both public and private to /home/sammy/.minio/certs  as below
```
home/sammy/.minio/certs$ ls -lrt
total 8
-rw-r--r-- 1 minio-user minio-user 806 Jun 17 03:58 public.crt
-rw------- 1 minio-user minio-user 241 Jun 17 03:58 private.key
```

**Reply by harshavardhana** [reply]
yes but you didn't put `public.crt` into the `/home/sammy/.minio/certs/CAs` folder - you need to do that for both sites to trust each other - so save them as two separate files such as `site1.crt` and `site2.crt`

**Reply by kittisuw** [reply]
let me try

**Reply by harshavardhana** [reply]
You should really read the docs https://docs.min.io/minio/baremetal/security/network-encryption/minio-tls.html#id2

**Reply by kittisuw** [reply]
As I try to copy public key of both servers to folder ./CAs/  of minio and mcli, first I test check info on server1 without --insecure the result is ok so I think It can communicate with public key. but when I use replicate command the error will be the same. I'm not quite sure on server2 should be installed mcli.(Server2 Install only minio)

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/53901743/174325144-a86a6e76-9f67-4808-b278-e78387ea8011.png">

**Reply by ravindk89** [reply]
If you ping one node to the other, what IP resolves? Make sure that from poc-minio-01 and poc-minio-02 they are seeing the correct hosts as listed in the CA. 

I've seen this issue elsewhere and it was a question of getting the CAs onto both servers, where the CA had to have the correct set of hostnames. since you are only using IP san, make sure that the IPs match exactly and are consistent for both hosts.

**Reply by kittisuw** [reply]
Really cool I just add trust hostname of both server and regenerate cert, I think the problem is when I using replicate it should trust cert by hostname. Thank for your help and support.
```shell
sudo certgen -host poc-minio-01,poc-minio-02,40.65.137.16,10.50.128.8,52.148.71.42,10.50.128.9
```
<img width="1436" alt="image" src="https://user-images.githubusercontent.com/53901743/174410692-511532fe-7740-42fd-88f3-1077008ef2b2.png">

**Reply by kittisuw** [reply]
Before replicate I can crete bucket without ERROR but After replicate success why I see error below(But the bucket was created successfully)
```bash
mcli: <ERROR> Unable to make bucket `poc-minio-01/bucket2`. Your previous request to create the named bucket succeeded and you already own it.
```
<img width="731" alt="image" src="https://user-images.githubusercontent.com/53901743/174427219-a1d0d679-991c-4784-9db9-d45d0142a33d.png">

**Reply by ravindk89** [reply]
What is the output of `mcli --insecure ls poc-minio-01` and `mcli --insecure ls poc-minio-02` ? 

The error seems to indicate that `bucket2` already existed - if you created it on `poc-minio-02` after configure site replication, it would have sync'd already to `poc-minio-01`. 

It _could_ be some sort of false error, but we'd need more data to confirm that. Lets start by validating your current existing buckets.

From there you can do in two shells:

```shell
# Shell 1

mcli admin --insecure trace poc-minio-01
```

```shell
# Shell 2

mcli mb --insecure poc-minio-01/testbucket
```

This will show us what happens on `poc-minio-01` when you create the bucket.

You can then repeat this, but running trace against `poc-minio-02` and creating a new bucket to see what happens via replication. This data might be helpful, assuming you see the issue consistently.

**Reply by harshavardhana** [reply]
The error that you see is expected when one of the site is down, as it returns for first attempt and upon a subsequent retry fails with the bucket already exists.

**Reply by kittisuw** [reply]
Thank for support I was fixed by non-ssl on minio server and using SSL on Nginx. Here is my step by step https://github.com/kittisuw/minio-replicate

**Reply by rasos** [reply]
@kittisuw thank you for sharing your setup. However, you are now replicating not encrypted and share accessKey and accessSecret without TLS. I did not have luck to replicate via nginx proxy. The server does not respond correctly on either console or API port when using nginx: 

> mc: <ERROR> Unable to get service status. S3 API Requests must be made to API port.
